streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableInnerJoin.java

```diff
@@ -87,22 +91,27 @@ public void process(final K key, final Change<V1> change) {
             }
 
             R newValue = null;
+            final long resultTimestamp;
             R oldValue = null;
 
-            final V2 value2 = valueGetter.get(key);
-            if (value2 == null) {
+            final ValueAndTimestamp<V2> valueAndTimestampRight = valueGetter.get(key);
+            final V2 valueRight = getValueOrNull(valueAndTimestampRight);
+            if (valueRight == null) {
                 return;
             }
 
             if (change.newValue != null) {
-                newValue = joiner.apply(change.newValue, value2);
+                newValue = joiner.apply(change.newValue, valueRight);
+                resultTimestamp = Math.max(context().timestamp(), valueAndTimestampRight.timestamp());
+            } else {
+                resultTimestamp = context().timestamp();
```

Comment:
Well. If time is properly synchronized, the example you describe should never happen, because `LEFT a:null (1)` should have been processed before `RIGHT A:y (2)`. Of course, the scenario you describe is possible if we enforce processing after `max.idle.time.ms` passed.

Consider the right hand side timestamp is somewhat weird to me though, because the right hand side record has nothing to do with the result. It's an inner join, and the left hand side tombstone determines that the result is deleted at tombstone event time (at least that is my current reasoning), and thus the result tombstone should have the same timestamp as the input record tombstone. The out-of-order problem is not introduced in this operator, but the operator rather propagates the out-of-order issue downstream, what seems "correct" to me.

Happy to be convinced otherwise, however, I don't think that the join has to "fix" out-of-order for this case but should rather propagate it downstream.

Reply:
I agree with John that we don't want the ts of the result to go backwards.

> Consider the right hand side timestamp is somewhat weird to me though, because the right hand side record has nothing to do with the result. 

I'm not familiar with this code, but conceptually it seems to make sense that the result tombstone has a ts of max(last_result.ts, left.ts).

Reply:
I am not convinced yet. Let's change the example a little bit and do:
```
observe RIGHT A:y (ts 20)
observe LEFT A:x (ts 0)
result A:(x,y) (ts 20)
observe LEFT A:null (ts 10)
result A:null (ts 10) // this is the tombstone in question -- alternative ts=20
observe RIGHT A:y2 (ts 15)
observe LEFT A:x2 (ts 17)
result A:(x1,x2) (ts 17)
```

With no out-of-order records, processing would be:
 - `LEFT A:x (ts 0)`
 - `LEFT A:null (ts 10)` // left table empty
 - `RIGHT A:y2 (ts 15)` -> result `A:(x,y2) (ts 15)`
 - `LEFT A:x2 (ts 17)` -> result `A:(x2,y2) (ts 17)`
 - `RIGHT A:y (ts 20)` -> result `A:(x2,y) (ts 20)`

For the processing with out-of-order data: If the result tombstone would be `A:null (ts 20)` it would semantically imply, that this tombstone deletes the final result  `A:(x1,x2) (ts 17)` (or actually correct result `A:(x2,y) (ts 20)`. In any case, the result is semantically empty. However, this sounds incorrect to me. The tombstone is rather a correction of the result table (that is at ts 20 already), but the correction _happens_ at event time 10. If there is no other update to the result KTable, the correction at ts=10 applies to the table at timestamp 20, too. However, if there is another update in between (as in our case, a new result at timestamp 17), the correction applies only from ts=10 to ts=17.

However, we can still see, that with out-of-order records, independent of the tombstone discussion, the compute result is not correct... However, fixing this is another mayor PR...

Bottom line: we are not able to handle out-of-order data gracefully for table-table-joins atm, and this PR won't fix it. We need to figure this out as follow up. However, I still think that using the input-tombstone-ts as output-tombston-ts is correct for this case.

Thoughts?

Reply:
Humm...

Taking a subset of your example:

```
observe RIGHT A:y (ts 20)
observe LEFT A:x (ts 0)
result A:(x,y) (ts 20)
observe LEFT A:null (ts 10)
result A:null (ts 10) // this is the tombstone in question -- alternative ts=20
```

If the system goes with ts of 10 for the result tombstone, would not downstream consumers be right to ignore this update as its an update to a table entry with a timestamp less that the current value, i.e. they'd keep `result A:(x,y) (ts 20)` and ignore `result A:null (ts 10)` as its older.  Therefore their end state would be a table with `result A:null (ts 10)`.

This is a table-table inner join, right?  Does this not provide us with some controls / guarantees over update ordering/timestamps? Is it not the case that the source tables themselves would  


```
observe RIGHT A:y (ts 20)
observe LEFT A:x (ts 0)
result A:(x,y) (ts 20)
observe LEFT A:null (ts 10)
result A:null (ts 10) // this is the tombstone in question -- alternative ts=20
observe RIGHT A:y2 (ts 15)
observe LEFT A:x2 (ts 17)
result A:(x1,x2) (ts 17)
```

| LEFT | RIGHT |
|---|----|
|  |  |
|  |  |
|  | A:y (ts 20) |
|  |  |
|  |  |

Reply:
Ack, my example was incorrect.... `RIGHT A:y2 (ts 15) // no output as left table empty` is correct.

>  I thought the table itself would ensure such out-of-order output from the table was not possible.

Out-of-order updates are indeed possible. Before KIP-258, updates to KTable (eg, reading from an input topic) are applied in offset order. Because timestamp information was not stored, it was not even possible to detect out-of-order updates and to discard them. With KIP-258, we had the discussion to now discard out-of-order updates of source-KTables, however, we decided to _not_ do this, due to backward compatibility concerns. We need to revisit this as a follow up...

About your first argument. I see your point. However, because the downstream KTable would apply the update (independent of the timestamp) as-of now, the incorrect result would be deleted. I agree, that we need to revisit this behavior, but it's out-of-scope for now.

To really fix the issue, I believe we actually need to distinguish between a "delete" and a "revert"...

Reply:
Seems to me that out-of-order messages on a single key should not be accepted on a table.  On a stream of data it is probably more common to have out-of-order updates than not, due to the distributed nature of modern systems. But we should require more strict ordering for a topic that represents a table.

Reply:
While I agree, the problem is that the current behavior is to accept out-of-order updates to `KTables` when reading from a topic. This PR does not change this for backward compatibility reasons (we only log a WARN in `KTableSource`). IMHO, we need a KIP and an opt-in model to make this change (and I agree, that we should do this).

Hence, we somehow need to handle the out-of-order case in this PR... I am not 100% sure if the current code, or the proposal is the better approach to be honest. On the other hand, we don't have a public contract about timestamp propagation yet, and we might need to revisit this again later. Thus, whatever we decide, the impact should be minor.

While I see the point about "not going back in time" and I also see the argument about 

> If the system goes with ts of 10 for the result tombstone, would not downstream consumers be right to ignore this update as its an update to a table entry with a timestamp less that the current value, i.e. they'd keep result A:(x,y) (ts 20) and ignore result A:null (ts 10) as its older. Therefore their end state would be a table with result A:null (ts 10).

I still lean slightly to the current behavior of the PR. But if you insist to change it, I am also ok, as I think we need to change it again later on anyway...

Reply:
Woah, awesome discussion.

Yeah, the "not going back in time" thing for KTables is not currently implemented, but it does seem like it's a precondition to having well-defined time semantics. So, we should consider join semantics assuming "correct" inputs from the source KTables. I.e., we should assume no out-of-order updates on the inputs to this join. In that case, then Andy's example still works , and we still get the correct final output.

The "incorrect" result from Matthias's example is an artifact of the "incorrect" KTable semantics. It doesn't seem like we should be "sloppy" with joins just because we already decided to be "sloppy" with Tables. There's no easy solution to the table semantics, but there is for the joins.

Another way to look at it is, if we "fix" the join time semantics, then all we have to do to get a sound system is to fix the table semantics. But if we leave the join semantics as is in this PR, then we have to fix _both_ tables and joins later on.

Does that seem to add up?

Reply:
It does make sense.

One thing to note is, that we use the input tombstone ts for output ts in other joins, too. I think we should at least be consistent. On the other joins, nobody raised concerns about using the tombstone timestamp, hence, it seem to be fine.

Might also be good if you look into the corresponding tests, and comment on the line for which you think a different timestamp (and which one) would be better?

Reply:
While I also feel that the ts should not go backwards since we'll follow up work to fix out-of-order data, I'd be ok with the PR as is.

Reply:
I'm logging off to do bedtime stories I'm afraid, so don't have time to annotate tests.

I'm with John when it comes to making this as correct and as inline with how we want the system to work long term, if we can.

Though I can see benefit in consistency with other join types, I think that being consistently wrong is probably worse than being inconsistent. ;)

I think a good litmus is whether someone can come along and materialize the output into some traditional database using the key and the timestamp and end up with the right result.  This is likely a fairly common thing people want to do, so should work as expected.

Reply:
Well. It's easy to change in all joins... all part of this PR. Let's see what @bbejeck thinks. I am 50:50 and it seems that @vvcephei and @big-andy-coates have strong opinion and good arguments to change it. I am inclined to update this atm.

\cc @guozhangwang If you want to chime in.